src/debugAdapter/goDebug.ts

```diff
@@ -479,65 +479,63 @@ class Delve {
 		});
 	}
 
-	callPromise<T>(command: string, args: any[]): Thenable<T> {
+	public callPromise<T>(command: string, args: any[]): Thenable<T> {
 		return new Promise<T>((resolve, reject) => {
 			this.connection.then(conn => {
-				conn.call<T>('RPCServer.' + command, args, (err, res) => {
-					if (err) return reject(err);
-					resolve(res);
+				conn.call<T>(`RPCServer.${command}`, args, (err, res) => {
+					return err ? reject(err) : resolve(res);
 				});
 			}, err => {
 				reject(err);
 			});
 		});
 	}
 
-	close(): Thenable<void> {
+	public close(): Thenable<void> {
 		log('HaltRequest');
 
-		return new Promise(resolve => {
-			let timeoutToken: NodeJS.Timer;
-			if (this.debugProcess) {
-				timeoutToken = setTimeout(() => {
-					log('Killing debug process manually as we could not halt and detach delve in time');
-					killTree(this.debugProcess.pid);
-					resolve();
-				}, 1000);
+		return new Promise(async resolve => {
+			const timeoutToken: NodeJS.Timer = this.debugProcess && setTimeout(() => {
+				log('Killing debug process manually as we could not halt and detach delve in time');
+				killTree(this.debugProcess.pid);
+				resolve();
+			}, 1000);
+
+			try {
+				await this.callPromise('Command', [{ name: 'halt' }]);
+			}
+			catch (err) {
+				const errMsg = err ? err.toString() : '';
+				log(`Failed to halt - ${errMsg}`);
+			}
+
+			if (timeoutToken) {
+				clearTimeout(timeoutToken);
 			}
 
-			this.callPromise('Command', [{ name: 'halt' }]).then(() => {
-				if (timeoutToken) {
-					clearTimeout(timeoutToken);
+			log('HaltResponse');
+			if (!this.debugProcess) {
+				log('RestartRequest');
+				try {
+					await this.callPromise('Restart', this.isApiV1 ? [] : [{ position: '', resetArgs: false, newArgs: [] }]);
 				}
-				log('HaltResponse');
-				if (!this.debugProcess) {
-					log('RestartRequest');
-					return this.callPromise('Restart', this.isApiV1 ? [] : [{ position: '', resetArgs: false, newArgs: [] }])
-						.then(null, err => {
-							log('RestartResponse');
-							logError(`Failed to restart - ${(err || '').toString()}`);
-						})
-						.then(() => resolve());
-				} else {
-					log('DetachRequest');
-					return this.callPromise('Detach', [this.isApiV1 ? true : { Kill: true }])
-						.then(null, err => {
-							log('DetachResponse');
-							logError(`Killing debug process manually as we failed to detach - ${(err || '').toString()}`);
-							killTree(this.debugProcess.pid);
-						})
-						.then(() => resolve());
+				catch (err) {
+					log('RestartResponse');
+					logError(`Failed to restart - ${(err || '').toString()}`);
 				}
-			}, err => {
-				const errMsg = err ? err.toString() : '';
-				log('Failed to halt - ' + errMsg.toString());
-				if (errMsg.endsWith('has exited with status 0')) {
```

Comment:
in the previous logic it seems we continued with `restart/detach` if it failed for any other reason than that the debugged processed finished. It seems that we would want exactly the opposite, i.e. continue with detach if the proc already exited.  

Question remains if we want to still continue (as we do today ) if halt fails from another reason than that of the debugged process being finished.

Reply:
When halt fails can we tell if it was due to the target process exited?

Reply:
well kind of, based on the message, like the `if` that was before these changes. So thats pretty much what im asking, do we want to detach only when the target exited, or always when halt fails ( like I mentioned, today we do the exact opposite, we always detach EXCEPT if the target process exited ).

 TBF, not sure why we wouldn't try to detach always if halt fails, I assume there is no point in continuing to be attached, but maybe @ramya-rao-a  or you have more context if there is a reason not to.

Reply:
Apologies for the late response.

To summarize, what we are trying to debate here is that when halt fails, if we should restart/detach only if the failure was due to target program exit or for all failures, correct?

If the target program is still running, and halt fails, there is actually no point in restart/detach as these requests would fail as well. If the target program is still running, delve needs to be halted in order to accept any further requests

Reply:
@ramya-rao-a  correct thats what we are saying. and to make sure we are on the same line, you are suggesting that : 
 - if halt succes -> then detach
 - if halt fails because target has exited -> then detach 
  - if halt fails because of whatever other reason -> doesn't really matter if we detach since it will fail but preferably not so we dont do an extra unnecessary operation

Reply:
Yes